genesis/engine/couplers/sap_coupler.py

```diff
@@ -0,0 +1,1971 @@
+        self.options = options
+        self.rigid_solver = self.sim.rigid_solver
+        self.fem_solver = self.sim.fem_solver
+        self._n_sap_iterations = options.n_sap_iterations
+        self._sap_threshold = options.sap_threshold
+        self._n_pcg_iterations = options.n_pcg_iterations
+        self._pcg_threshold = options.pcg_threshold
+        self._n_linesearch_iterations = options.n_linesearch_iterations
+        self._linesearch_c = options.linesearch_c
+        self._linesearch_tau = options.linesearch_tau
+        self._use_exact_linesearch = options.use_exact_linesearch
+        self._fem_floor_tet = options.fem_floor_tet
+        self._fem_floor_vert = options.fem_floor_vert
+        self._fem_self_tet = options.fem_self_tet
+
+    # ------------------------------------------------------------------------------------
+    # --------------------------------- Initialization -----------------------------------
+    # ------------------------------------------------------------------------------------
+
+    def build(self) -> None:
+        self._B = self.sim._B
+        self._rigid_fem = self.rigid_solver.is_active() and self.fem_solver.is_active() and self.options.rigid_fem
+        self.contacts = []
+        self.init_bvh()
+
+        if self.fem_solver.is_active():
+            if self.fem_solver._use_implicit_solver is False:
+                gs.raise_exception(
+                    "SAPCoupler requires FEM to use implicit solver. "
+                    "Please set `use_implicit_solver=True` in FEM options."
+                )
+            if self._fem_floor_tet or self._fem_self_tet:
+                # Hydroelastic
+                self.init_hydroelastic_fem_fields_and_info()
+
+            if self._fem_floor_tet:
+                self.fem_floor_tet_contact = FEMFloorTetContact(self.sim)
+                self.contacts.append(self.fem_floor_tet_contact)
+
+            if self._fem_floor_vert:
+                self.fem_floor_vert_contact = FEMFloorVertContact(self.sim)
+                self.contacts.append(self.fem_floor_vert_contact)
+
+            if self._fem_self_tet:
+                self.fem_self_tet_contact = FEMSelfTetContact(self.sim)
+                self.contacts.append(self.fem_self_tet_contact)
+
+        self.init_sap_fields()
+        self.init_pcg_fields()
+        self.init_linesearch_fields()
+
+    def reset(self, envs_idx=None):
+        pass
+
+    def init_hydroelastic_fem_fields_and_info(self):
+        fem_solver = self.fem_solver
+        self.fem_pressure = ti.field(gs.ti_float, shape=(fem_solver.n_vertices))
+        fem_pressure_np = np.concatenate([fem_entity.pressure_field_np for fem_entity in fem_solver.entities])
+        self.fem_pressure.from_numpy(fem_pressure_np)
+        self.fem_pressure_gradient = ti.field(gs.ti_vec3, shape=(fem_solver._B, fem_solver.n_elements))
+
+        # Lookup table for marching tetrahedra edges
+        kMarchingTetsEdgeTable_np = np.array(
```

Comment:
Make this a global constant outside the class (as nested tuples to make it immutable):
```python
MARCHING_TETS_EDGE_TABLE = (
    (-1, -1, -1, -1),
    [...]
)
```